File: training/3.2/sweet_butter/butter.py
"""
PROG: butter
LANG: PYTHON3

"""
import sys
import queue

with open('butter.in', 'r') as fin:
  lines = fin.readlines()

N, P, C = map(int, lines[0].split())

pastures = []
for i in range(N):
    pastures.append(int(lines[i+1])-1)

d = [[sys.maxsize for _ in range(P)] for _ in range(P)]

edges = {}
for i in range(C):
    p1, p2, lenP = map(int, lines[N+1+i].split())
    p1n = p1 - 1
    p2n = p2 -1
    if p1n not in edges.keys():
        edges[p1n] = []
    edges[p1n].append([p2n, lenP])
    if p2n not in edges.keys():
        edges[p2n] = []
    edges[p2n].append([p1n, lenP])
    
for i in range(P):
    d[i][i] = 0

def spfa(d, edges, u):
    q = queue.Queue()
    q.put(u)
    visited = [False] * P
    #print("u=", u)
    while not q.empty():
        cur = q.get()
        #print("cur=", cur)
        for edge in edges[cur]:
            if d[u][edge[0]] > d[u][cur] + edge[1]:
                d[u][edge[0]] = d[u][cur] + edge[1]
                if visited[edge[0]] == False:
                    q.put(edge[0])
                    #print("pushing=", edge[0])
                    visited[edge[0]] = True
        visited[cur] = False
        
"""
for k in range(P):
    for i in range(P):
        for j in range(P):
            if i != j and connections[i][k] != 0 and connections[j][k] != 0 and (connections[i][j] == 0 or connections[i][j] > connections[i][k] + connections[j][k]):
                connections[i][j] = connections[i][k] + connections[j][k]
"""

for i in range(P):
    spfa(d, edges, i)

result = sys.maxsize
for i in range(P):
    totalLen = 0
    for pasture in pastures:
        totalLen += d[i][pasture]
    if result > totalLen:
        result = totalLen
        #print("i=", i)

with open('butter.out', 'w') as fout:
    fout.write(str(result) + '\n')

File: training/3.2/sweet_butter/butter.cpp
/*
PROG: butter
LANG: C++
*/

#include <iostream>
#include <fstream>
#include <vector>
#include <queue>

using namespace std;

#define MYMAX (1 << 30)
#define PASTURES 800
#define COWS 500

struct Edge {
    int pasture;
    int distance;
    Edge(int p, int d) : pasture(p), distance(d) {}
};

int d[PASTURES][PASTURES];
int pastures[COWS];
vector<Edge> edges[PASTURES];

void spfa(int u, int P, const vector<Edge> edges[PASTURES], int d[PASTURES][PASTURES])
{
    queue<int> q;
    q.push(u);
    bool visited[PASTURES];
    //cout << "u=" << u << endl;
    for (int i = 0; i < P; i++)
        visited[i] = false;
    while(!q.empty())
    {
        int cur = q.front();
        //cout << "cur=" << cur << endl;
        q.pop();
        for (int i = 0; i < edges[cur].size(); ++i)
        {
            int v = edges[cur][i].pasture;
            int distance = edges[cur][i].distance;
            //if (cur == 3 && v == 6)
                //cout << "d[0][6]=" << d[0][6] << ", d[0][3]=" << d[0][3] << ", w[3][6]=" << w[3][6] << ", visited[6]=" << visited[6] << endl;
            if (d[u][v] > d[u][cur] + distance)
            {
                d[u][v] = d[u][cur] + distance;

                if (visited[v] == false)
                {
                    q.push(v);
                    //cout << "pushing=" << v << endl;
                    visited[v] = true;
                }
            }
        }
        visited[cur] = false;
    }
}

int main()
{
    ifstream fin("butter.in");
    ofstream fout("butter.out");

    int N, P, C;
    fin >> N >> P >> C;

    for (int i = 0; i < N; ++i)
    {
        fin >> pastures[i];
        pastures[i] -= 1;
    }

    for (int i = 0; i < P; ++i)
        for (int j = 0; j < P; ++j)
        {
            d[i][j] = MYMAX;
        }

    for (int i = 0; i < C; ++i)
    {
        int start, end, size;
        fin >> start >> end >> size;
        
        edges[start-1].push_back(Edge(end-1, size));
        edges[end-1].push_back(Edge(start-1, size));
    }

    for (int i = 0; i < P; ++i)
    {
        d[i][i] = 0;
    }

    for (int i=0; i < P; ++i)
    {
        spfa(i, P, edges, d);
    }
    int result = MYMAX;
    for (int i = 0; i < P; ++i)
    {
        int totalLen = 0;
        for (int j = 0; j < N; ++j)
            totalLen += d[i][pastures[j]];
        if (result > totalLen)
        {
            result = totalLen;
            //cout << "i=" << i << endl;
        }
    }

    fout << result << endl;
}